pkg/virtctl/create/vm.go

```diff
@@ -0,0 +1,623 @@
+		return flagErr(InstancetypeFlag, "invalid instancetype kind \"%s\", supported values are: %s, %s", kind, instancetype.SingularResourceName, instancetype.ClusterSingularResourceName)
+	}
+
+	// If kind is empty we rely on the vm-mutator to fill in the default value VirtualMachineClusterInstancetype
+	vm.Spec.Instancetype = &v1.InstancetypeMatcher{
+		Name: name,
+		Kind: kind,
+	}
+
+	return nil
+}
+
+func withPreference(c *createVM, vm *v1.VirtualMachine) error {
+	kind, name, err := splitPrefixedName(c.preference)
+	if err != nil {
+		return flagErr(PreferenceFlag, "%w", err)
+	}
+
+	if kind != "" && kind != instancetype.SingularPreferenceResourceName && kind != instancetype.ClusterSingularPreferenceResourceName {
+		return flagErr(InstancetypeFlag, "invalid preference kind \"%s\", supported values are: %s, %s", kind, instancetype.SingularPreferenceResourceName, instancetype.ClusterSingularPreferenceResourceName)
+	}
+
+	// If kind is empty we rely on the vm-mutator to fill in the default value VirtualMachineClusterPreference
+	vm.Spec.Preference = &v1.PreferenceMatcher{
+		Name: name,
+		Kind: kind,
+	}
+
+	return nil
+}
+
+func withContainerdiskVolume(c *createVM, vm *v1.VirtualMachine) error {
+	for i, containerdiskVol := range c.containerdiskVolumes {
+		vol := containerdiskVolume{}
+		err := mapParams(ContainerdiskVolumeFlag, containerdiskVol, &vol)
+		if err != nil {
+			return err
+		}
+
+		if vol.Source == "" {
+			return flagErr(ContainerdiskVolumeFlag, "src must be specified")
+		}
+
+		if !strings.HasPrefix(vol.Source, "docker://") {
+			vol.Source = fmt.Sprintf("docker://%s", vol.Source)
+		}
+		if vol.Name == "" {
+			vol.Name = fmt.Sprintf("%s-containerdisk-%d", vm.Name, i)
+		}
+
+		if err := checkVolumeExists(ContainerdiskVolumeFlag, vm.Spec.Template.Spec.Volumes, vol.Name); err != nil {
+			return err
+		}
+
+		vm.Spec.Template.Spec.Volumes = append(vm.Spec.Template.Spec.Volumes, v1.Volume{
+			Name: vol.Name,
+			VolumeSource: v1.VolumeSource{
+				ContainerDisk: &v1.ContainerDiskSource{
+					Image: vol.Source,
+				},
+			},
+		})
+	}
+
+	return nil
+}
+
+func withDataSourceVolume(c *createVM, vm *v1.VirtualMachine) error {
+	for _, dataSourceVol := range c.dataSourceVolumes {
+		vol := cloneVolume{}
+		err := mapParams(DataSourceVolumeFlag, dataSourceVol, &vol)
+		if err != nil {
+			return err
+		}
+
+		if vol.Source == "" {
+			return flagErr(DataSourceVolumeFlag, "src must be specified")
+		}
+
+		namespace, name, err := splitPrefixedName(vol.Source)
+		if err != nil {
+			return flagErr(DataSourceVolumeFlag, "src invalid: %w", err)
+		}
+
+		if vol.Name == "" {
+			vol.Name = fmt.Sprintf("%s-ds-%s", vm.Name, name)
+		}
+
+		if err := checkDvtExists(DataSourceVolumeFlag, vm.Spec.DataVolumeTemplates, vol.Name); err != nil {
+			return err
+		}
+		if err := checkVolumeExists(DataSourceVolumeFlag, vm.Spec.Template.Spec.Volumes, vol.Name); err != nil {
+			return err
+		}
```

Comment:
We only need to check for duplicate volumes right? There shouldn't be a case where this command defines a DVT that doesn't have the same name as a volume?

Reply:
Actually this isn't the end of the world and might catch issues in the future if we somehow allow users to provide different DVT names to volumes.

Reply:
I changed it to only check for duplicate volumes for now.